server/src/main/java/org/elasticsearch/index/mapper/MapperService.java

```diff
@@ -590,6 +584,10 @@ public MappedFieldType fieldType(String fullName) {
         return fieldTypes.get(fullName);
     }
 
+    public Mapper fieldMapper(String fieldName) {
```

Comment:
This is unfortunately necessary as a shim to get round the ParentJoinFieldMapper issue above.  Hopefully removed asap.

Reply:
Could we avoid adding this new method and call this inline in `ParentJoinFieldMapper`? All the methods like `documentMapper()` are public, so it seems do-able.

Reply:
++ have reworked the mocking (which sent me down a surprisingly deep rabbit hole, but I think it's fixed now!)